### ISSUE TITLE
Question about reward sum in TD loss

### ISSUE THREAD

**Issue by oo1232** [ColinQiyangLi/qc#11]
Hi Colin:

Thanks for your excellent work! 

I notice that in the TD loss section of your paper, you have summed the rewards, but in the actual code implementation, only the final reward is utilized. Could you please explain the rationale behind this design choice?

<img width="1128" height="147" alt="Image" src="https://github.com/user-attachments/assets/e5cc51ed-ea57-43bf-850e-3499f7e16872" />

        target_q = batch['rewards'][..., -1] + (self.config['discount'] ** self.config["horizon_length"]) * batch['masks'][..., -1] * next_q

Thanks a lot!

**Comment by ColinQiyangLi** [comment]
Hi, thank you for your kind word! 

Regarding your question -- `batch['rewards'][..., -1]` actually stores the h-step reward (I should have really called it "n_step_rewards" or something) as implemented here: https://github.com/ColinQiyangLi/qc/blob/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee/utils/datasets.py#L126

Hope this clarifies it!

**Comment by oo1232** [comment]
Hi Colin:
Thank you very much for your response; it addresses my previous concern. I would like to follow up with an additional question. In the current dataset construction method, is it possible for a sample's sequence to contain data segments originating from two distinct data-part? If so, how is the reward function designed for this specific scenario?

**Comment by ColinQiyangLi** [comment]
No problem! For your follow-up question, every segment is associated with a `valid` flag. It is only true when the segment comes from a single trajectory. If the segment originates from two trajectories, the valid flag is set to false (https://github.com/ColinQiyangLi/qc/blob/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee/utils/datasets.py#L129) and the critic loss is turned off for these trajectories as well (https://github.com/ColinQiyangLi/qc/blob/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee/agents/acfql.py#L45). In hindsight, a cleaner way to implement this is probably to just avoid sampling these trajectory segments directly in the data loader side.

**Comment by oo1232** [comment]
Thank you very much for your response; I now understand your current configuration and agree with its validity. However, I have an additional question concerning task types such as those in RoboMimic, where the rewards are typically sparse, and a non-negative reward is only granted upon successful completion. Under these conditions, the Critic can only learn effective success information when the prediction horizon precisely aligns with the end of a successful trajectory. Could you comment on whether this specific setup might result in a sub-optimal learning efficiency for the Critic network?

<img width="774" height="335" alt="Image" src="https://github.com/user-attachments/assets/860a20e3-d732-43ec-8a1e-3a2647b9a5ca" />

**Comment by ColinQiyangLi** [comment]
No problem! Your understanding is correct. I admit it is a bit non-ideal and could potentially hurt learning efficiency, but I don't expect it to be an issue for the benchmark tasks that we tested on. The sparsity issue should be similar to that of standard 1-step TD (only receive positive signal at the end of the trajectory). But perhaps figuring out a way to make use of truncated segments well could make QC's performance even better! 

Something that I did not get to it when I was working on this project but could be a way to handle this boundary condition cleanly --- 

First train the action chunking policy in a way that could predict the full chunk from any prefix actions (e.g., following the behavior distribution or something). This could be done by any policy model that supports autoregressive sampling.

$\pi(a_{t+h_p:t+h} \mid a_{t:t+h_p}, s_t)$

So for any trajectory segment that is truncated **because of success** (or in general any termination that is not due to episode length truncation), we could just use this model to first complete the action chunk and then regress the chunked Q-function towards the truncated n-step reward in the segment:

$Q(s_t, a_{t:t+h}) \leftarrow \sum_{t'=t}^{t+h_p-1} \gamma^{t'-t} r_{t'}, $
where $a_{t:t+h} = [a_{t:t+h_p}, a_{t+h_p: t+h}]$ with $a_{t+h_p: t+h} \sim \pi(a_{t+h_p:t+h} \mid a_{t:t+h_p}, s_t)$

$h_p$ is the length of the trajectory segment after truncation.

Any trajectory segment truncated **not due to success** (or in general due to episode length truncation) still needs to be marked as invalid though.

Hope this helps!

**Comment by oo1232** [comment]
Thank you very much for your detailed response. I would like to express my appreciation once again for your outstanding work!

**Comment by hosh0425** [comment]
> No problem! For your follow-up question, every segment is associated with a `valid` flag. It is only true when the segment comes from a single trajectory. If the segment originates from two trajectories, the valid flag is set to false (
> 
> [qc/utils/datasets.py](https://github.com/ColinQiyangLi/qc/blob/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee/utils/datasets.py#L129)
> 
> Line 129 in [e03c80d](/ColinQiyangLi/qc/commit/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee)
>  valid[:, i] = 1.0 - terminals[:, i-1] 
> ) and the critic loss is turned off for these trajectories as well (
> 
> [qc/agents/acfql.py](https://github.com/ColinQiyangLi/qc/blob/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee/agents/acfql.py#L45)
> 
> Line 45 in [e03c80d](/ColinQiyangLi/qc/commit/e03c80dab5aa5e85c2cccdf8fbb153091bfa62ee)
>  critic_loss = (jnp.square(q - target_q) * batch['valid'][..., -1]).mean() 
> ). In hindsight, a cleaner way to implement this is probably to just avoid sampling these trajectory segments directly in the data loader side.

does not this mean that last `H=chunk_value` transitions of each episode do NOT contribute to the critic updates ever? Is not the implementation in DQC repo a better approach (i.e. affecting the effective horizon in TD targets)?
